### PR TITLE
initial worker skeleton

### PR DESCRIPTION
**Pull request by anuejn**

### CODE REVIEW

File: worker/src/torchaudio_align.py [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+    interpolate_method: str = "nearest",
+    start_from_previous: bool = True,
+) -> :
+    """
+    Force align phoneme recognition predictions to known transcription
+
+    Parameters
+    ----------
+    transcript: Iterator[dict]
+        The Whisper model instance
```
Comment (rroohhh) [inline worker/src/torchaudio_align.py:105]: Whisper model instance?
Why is a Whisper model instance a `Iterator[dict]`

File: worker/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+description = "a transcription worker for the transcribee open-source transcription platform"
+authors = [
+    {name = "anuejn", email = "jarohabiger@googlemail.com"},
+    {name = "pajowu", email = "git@ca.pajowu.de"},
+    {name = "phlmn", email = "info@philipp-mandler.com"},
+    {name = "vup", email = "robin.ole.heinemann+transcribee@gmail.com"},
+]
+
+dependencies = [
+    "transcribee-proto @ file:///${PROJECT_ROOT}/../proto",
```
Comment (rroohhh) [inline worker/pyproject.toml:13]: Missing from PR.

File: worker/src/torchaudio_align.py (hunk 1 above)
Comment (anuejn) [reply worker/src/torchaudio_align.py:105]: good point, this is wrong

**Comment by anuejn** [comment]
I think I would be happy to merge this in this state :)

File: worker/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    "whispercpp @ git+https://github.com/transcribee/whispercpp",
+    "transformers>=4.26.1",
+    "torchaudio>=0.13.1",
+    "torch>=1.13.1",
+]
+requires-python = ">=3.10"
+readme = "README.md"
+license = {text = "AGPL-3.0"}
+
+[tool.pdm]
```
Comment (rroohhh) [inline worker/pyproject.toml:27]: This section is not needed

File: worker/transcribe_manual.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+from transcribee_proto.document import Document, Paragraph
+from transcribee_worker.torchaudio_align import align
+from transcribee_worker.util import load_audio
+from transcribee_worker.whisper_transcribe import transcribe
+
+logging.basicConfig(level=logging.INFO)
+
+if __name__ == "__main__":
+    if len(sys.argv) != 3:
```
Comment (rroohhh) [inline worker/transcribe_manual.py:18]: Lets use `argparse`

Comment (rroohhh) [reply worker/transcribe_manual.py:18]: (Also FYI `lang` is currently broken in the `whisper.cpp` python bindings

File: worker/transcribe_manual.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        print(f"usage:\n\t{sys.argv[0]} lang_code path/to/audio.mp3")
+        exit(-1)
+    else:
+        _, lang_code, audio_path = sys.argv
+
+    audio = load_audio(audio_path)
+    transcript = transcribe(audio, "tiny", lang_code)
```
Comment (rroohhh) [inline worker/transcribe_manual.py:25]: This is broken, `transcribe` is now async, etc

**Review by rroohhh — CHANGES_REQUESTED** [review]
Just had a rough look over it.

This also depends on #13 to be merged first.

Also note to self: We can use upstream `whispercpp` again, all changes needed by us are merged now. :tada:

File: proto/transcribee_proto/document.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import itertools
+from typing import Iterator, List, Literal, Union
+
+from pydantic import BaseModel
+
+
+class Atom(BaseModel):
+    text: str
+    # in ms
```
Comment (anuejn) [inline proto/transcribee_proto/document.py:9]: can we plz put these in the same line?

File: proto/transcribee_proto/document.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+class Paragraph(BaseModel):
+    type: Literal["paragraph"] = "paragraph"
+    speaker: str
+    children: List[Atom]
+
+    def text(self) -> str:
+        return "".join(a.text for a in self.children)
+
+    def start(self) -> Union[None, float]:
```
Comment (anuejn) [inline proto/transcribee_proto/document.py:25]: Can we use the new `float | None` syntax here?

File: worker/transcribe_manual.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+from transcribee_worker.torchaudio_align import align
+from transcribee_worker.util import load_audio
+from transcribee_worker.whisper_transcribe import transcribe
+
+logging.basicConfig(level=logging.INFO)
+
+
+async def main():
+    parser = argparse.ArgumentParser(
+        description="Manually execute the transcription worker"
```
Comment (anuejn) [inline worker/transcribe_manual.py:23]: not execute the worker but transcribe an audio file to a transcribee transcript

File: worker/transcribee_worker/config.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+from pathlib import Path
+
+# Samples per ms
+SAMPLE_RATE = 16
```
Comment (anuejn) [inline worker/transcribee_worker/config.py:4]: this is wiered. can we store it in HZ plz? then we can also remove the comment

File: worker/transcribee_worker/whisper_transcribe.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    ctx = api.Context.from_file(str(model_file))
+    ctx.reset_timings()
+    return ctx
+
+
+class TranscriptionWorkDoneToken:
+    pass
+
+
+# TODO(robin): this currently filters all special tokens
```
Comment (anuejn) [inline worker/transcribee_worker/whisper_transcribe.py:41]: can you remove your name there?

Comment (anuejn) [reply worker/transcribee_worker/whisper_transcribe.py:41]: & open an issue once this is merged?

File: worker/transcribee_worker/whisper_transcribe.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+def get_context(model_name: str) -> api.Context:
+    model_file = get_model_file(model_name)
+    logging.info(f"loading model {model_name}...")
+    ctx = api.Context.from_file(str(model_file))
+    ctx.reset_timings()
+    return ctx
+
+
+class TranscriptionWorkDoneToken:
```
Comment (anuejn) [inline worker/transcribee_worker/whisper_transcribe.py:37]: wouldnt it be more ideomatic if we used `StopIteration` here?

**Review by anuejn — COMMENTED** [review]
Looks nice :).
Just some random bikeshedding from my side.

File: worker/transcribee_worker/whisper_transcribe.py (hunk 10 above)
Comment (rroohhh) [reply worker/transcribee_worker/whisper_transcribe.py:41]: Why remove the name?
That way you save one `git blame` to figure out who to ask about a `TODO`.

Comment (anuejn) [reply worker/transcribee_worker/whisper_transcribe.py:41]: seems just not super nescessary to me here but no strong feelings on that

File: worker/transcribee_worker/whisper_transcribe.py (hunk 11 above)
Comment (rroohhh) [reply worker/transcribee_worker/whisper_transcribe.py:37]: I mean not really? This is not a `Iterator`, but I am fine using `StopIteration` if you prefer it anyways.

Comment (rroohhh) [reply worker/transcribee_worker/whisper_transcribe.py:37]: Well actually 
```
ERROR:asyncio:Exception in callback _chain_future.<locals>._set_state(<Future pendi...tasks.py:475]>, <Future at 0x...StopIteration>) at /nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/asyncio/futures.py:379
handle: <Handle _chain_future.<locals>._set_state(<Future pendi...tasks.py:475]>, <Future at 0x...StopIteration>) at /nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/asyncio/futures.py:379>
Traceback (most recent call last):
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/asyncio/futures.py", line 381, in _set_state
    _copy_future_state(other, future)
  File "/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/lib/python3.10/asyncio/futures.py", line 357, in _copy_future_state
    dest.set_exception(_convert_future_exc(exception))
TypeError: StopIteration interacts badly with generators and cannot be raised into a Future
```

**Pull request merged**